File: src/stack/lc_739_daily_temperatures/daily_temperatures.py
class Solution:
    #Time complex O(n)
    #Space complex O(n)
    def dailyTemperatures(self, temperatures: list[int]) -> list[int]:
        length = len(temperatures)
        res = [0] * length
        stack = []
        for i in range(len(temperatures)): #last value is always 0
            temp = temperatures[i]
            if not stack:
                stack.append((temp, i)) #store value and index
                continue
            while stack:
                if temp <= stack[-1][0]: # continue storing
                    break
                tempPre, idxPre = stack.pop()
                res[idxPre] = i - idxPre

            stack.append((temp, i))
        return res

File: src/stack/lc_739_daily_temperatures/daily_temperatures.java
import java.util.*;
class Solution {

    class Pair {
        int key;
        int idx;
        Pair(int key, int idx) {
            this.key = key;
            this.idx = idx;
        }

        public int getKey() {
            return this.key;
        }

        public int getIdx() {
            return this.idx;
        }
    }
    public int[] dailyTemperatures(int[] temperatures) {
        Stack<Pair> stack = new Stack<>();
        int[] res = new int[temperatures.length]; //init with all 0
        for(int i = 0; i < temperatures.length; i++) {
            int temperature = temperatures[i];
            if(stack.isEmpty()){
                stack.push(new Pair(temperature, i)); //add temp and index
                continue;
            }
            //check if current temp > temp at peek of stack
            while(!stack.isEmpty()){
                if (temperature <= stack.peek().getKey()) {
                    break;
                }
                //if temperature > previou one
                Pair pair = stack.pop();
                res[pair.getIdx()] = i - pair.getIdx(); //calculate num of days curr idx - prev idx
            }
            stack.push(new Pair(temperature, i));
        }
        return res;
    }
}